### PR TITLE
[react]: Remove deprecated+removed APIs

### PR DESCRIPTION
**Pull request by jgoz**
React v16 [has removed APIs deprecated in 15.5](https://reactjs.org/blog/2017/09/26/react-v16.0.html#packaging):
- `React.PropTypes` -> `prop-types`
- `React.createClass` -> `create-react-class`
- `React.DOM` -> `react-dom-factories`

This PR adds definitions for `create-react-class` and `react-dom-factories` and removes the deprecated APIs from the `react` definitions (while preserving the relevant interfaces and tests in the `react` package).

This is a breaking change, but that change was made in React 16 so the typings were actually wrong in 16.0.0 and could lead to unexpected runtime errors.

Tests were updated for a number of affected packages.

Note that this revealed existing errors in `redux-form` with TypeScript@next and `strictFunctionTypes` enabled. Since I am not familiar with that library, I chose to disable that option for the time being. One of its authors should investigate.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.

### CODE REVIEW

File: types/redux-form/v6/tsconfig.json
Comment (weswigham) [inline types/redux-form/v6/tsconfig.json]: We generally don't like going back on the strictness provided by a library - what became the issue here?

File: types/redux-form/tsconfig.json
Comment (weswigham) [inline types/redux-form/tsconfig.json]: We generally don't like going back on the strictness provided by a library - what became the issue here, too?

File: types/create-react-class/tslint.json
Comment (weswigham) [inline types/create-react-class/tslint.json]: It looks like there _is_ a DT header in the file; what's the issue here?

File: types/react-dom-factories/tslint.json
Comment (weswigham) [inline types/react-dom-factories/tslint.json]: Same here - the definition seems to have a dt header, so what's the issue here?

**Review by weswigham — CHANGES_REQUESTED** [review]
Also, calling
@johnnyreilly
@bbenezech
@pzavolinsky
@digiguru
@ericanderson
@morcerf
@tkrotoff
@DovydasNavickas
@onigoetz
@richseviora
for additional review. This being the react dts, more eyes are always good. 🐱

File: types/redux-form/tsconfig.json
Comment (jgoz) [reply types/redux-form/tsconfig.json]: I should have linked to the issue I filed: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20414

The above issue is present in master, but was only revealed in this PR due to the fact that I touched the `redux-form` tests. I disabled this option to get the tests passing. I'm not familiar with that particular library, and making changes to it seemed out of scope here.

File: types/create-react-class/tslint.json
Comment (jgoz) [reply types/create-react-class/tslint.json]: The issue is that there are no github links in the "Definitions by" field (same with the other package). I can add my own profile URL and re-enable this rule.

Comment (jgoz) [reply types/create-react-class/tslint.json]: Actually, this is an issue with putting _any_ non-GitHub URL in the "Definitions by" section (i.e., Microsoft, Asana, etc.). So I could re-enable this, but it would mean having only my name in that section. Any preference?

File: types/react-dom-factories/tslint.json
Comment (jgoz) [reply types/react-dom-factories/tslint.json]: Same issue as above.

File: types/create-react-class/tslint.json
Comment (weswigham) [reply types/create-react-class/tslint.json]: @Microsoft and @Asana are at least on github, so you can replace their corporate urls (which are, IMO, useless in this context) with the github org links.

File: types/redux-form/tsconfig.json
Comment (weswigham) [reply types/redux-form/tsconfig.json]: Welp. I guess disabling it for now is OK, then.

File: types/create-react-class/tslint.json
Comment (weswigham) [reply types/create-react-class/tslint.json]: Nvm, as per comment on other package, leave them as-is w/lint rule. Should update the rule to not fail on those domains.

Comment (jgoz) [reply types/create-react-class/tslint.json]: OK. Leaving as-is.

**Comment by jgoz** [comment]
@weswigham Any more issues on your end? Hoping the `Revision needed` label isn't putting off other reviewers.

**Comment by weswigham** [comment]
Nope, its fine from my perspective. I'll merge in a day or so if the other owners have no complaints.

**Comment by Flarna** [comment]
Please be aware that CI on master is currently broken for this module and all modules it depends on (e.a. node) because of the `"strictFunctionTypes": true` issue mentioned here. So if there are longer discussions for the other changes I would appreciate if the strictFunctionTypes part could be merge soon.

**Comment by weswigham** [comment]
I'll merge in the morning (~9 hours) to give the other maintainers an opportunity to comment if they so choose - the PR has only been out for 3 days and this is a popular package (and a somewhat large change). I wanna give it a fair shake, as I'd rather have more people looking at it than less. Good to see that others _are_ looking at it, though!

File: types/create-react-class/index.d.ts
Comment (andy-ms) [inline types/create-react-class/index.d.ts]: If you wrote these definitions, why are you crediting all these others?

File: types/jsnox/jsnox-tests.ts
Comment (andy-ms) [inline types/jsnox/jsnox-tests.ts]: I don't think there's any need to involve `create-react-class` here? It's better if types have fewer dependencies on each other.

File: types/create-react-class/index.d.ts
Comment (jgoz) [reply types/create-react-class/index.d.ts]: Well, I simply moved the definitions out of `react` and into these new packages. I thought it was more appropriate to keep the "Definitions by" in sync with `react`.

I can put my name on these new packages if you think that's more appropriate.

File: types/jsnox/jsnox-tests.ts
Comment (jgoz) [reply types/jsnox/jsnox-tests.ts]: I was trying to avoid changing the intent of the tests in dependent packages. I'm not familiar with `jsnox` (or most of these other libraries) so I wasn't sure if the `createClass` usage here was deliberate. Would you prefer to see instances like this changed to use component classes to avoid inter-dependencies?

File: types/create-react-class/index.d.ts
Comment (andy-ms) [reply types/create-react-class/index.d.ts]: Yes, please use your GH username here, you did create the package after all. :+1:

File: types/jsnox/jsnox-tests.ts
Comment (andy-ms) [reply types/jsnox/jsnox-tests.ts]: If it's possible, it would be better to have a pure-`react` test. Thanks.

**Comment by jgoz** [comment]
@andy-ms Updated based on your feedback, please have another look.

File: types/react/test/index.ts
Comment (andy-ms) [inline types/react/test/index.ts]: If `createClass` isn't a part of react anymore, no need to test it here?

Comment (jgoz) [reply types/react/test/index.ts]: Moved top-level tests to `create-react-class.tests.tsx`.

**Comment by andy-ms** [comment]
Thanks!

**Pull request merged**
